perception/autoware_probabilistic_occupancy_grid_map/lib/utils/utils_kernel.cu

```diff
@@ -0,0 +1,296 @@
+    return false;
+  }
+
+  mx = static_cast<int>(std::floor((wx - origin_x) * resolution_inv));
+  my = static_cast<int>(std::floor((wy - origin_y) * resolution_inv));
+
+  if (mx < size_x && my < size_y) {
+    return true;
+  }
+
+  return false;
+}
+
+__device__ void setCellValue(
+  float wx, float wy, float origin_x, float origin_y, float resolution_inv, int size_x, int size_y,
+  std::uint8_t value, std::uint8_t * costmap_tensor)
+{
+  unsigned int mx, my;
+  if (!worldToMap(wx, wy, mx, my, origin_x, origin_y, resolution_inv, size_x, size_y)) {
+    return;
+  }
+
+  costmap_tensor[my * size_x + mx] = value;
+}
+
+inline __device__ void bresenham2D(
+  unsigned int abs_da, unsigned int abs_db, int error_b, int offset_a, int offset_b,
+  unsigned int offset, unsigned int max_length, std::uint8_t cost, std::uint8_t * costmap_tensor)
+{
+  unsigned int end = min(max_length, abs_da);
+  for (unsigned int i = 0; i < end; ++i) {
+    costmap_tensor[offset] = cost;
+    offset += offset_a;
+    error_b += abs_db;
+    if ((unsigned int)error_b >= abs_da) {
+      offset += offset_b;
+      error_b -= abs_da;
+    }
+  }
+  costmap_tensor[offset] = cost;
+}
+
+/**
+ * @brief  Raytrace a line and apply some action at each step
+ * @param  at The action to take... a functor
+ * @param  x0 The starting x coordinate
+ * @param  y0 The starting y coordinate
+ * @param  x1 The ending x coordinate
+ * @param  y1 The ending y coordinate
+ * @param  max_length The maximum desired length of the segment...
+ * allows you to not go all the way to the endpoint
+ * @param  min_length The minimum desired length of the segment
+ */
+inline __device__ void raytraceLine(
+  unsigned int x0, unsigned int y0, unsigned int x1, unsigned int y1, unsigned int max_length,
+  unsigned int min_length, unsigned int size_x, std::uint8_t cost, std::uint8_t * costmap_tensor)
+{
+  int dx_full = x1 - x0;
+  int dy_full = y1 - y0;
+
+  // we need to chose how much to scale our dominant dimension,
+  // based on the maximum length of the line
+  float dist = sqrt((float)(dx_full * dx_full + dy_full * dy_full));  // hypot(dx_full, dy_full);
+  if (dist < min_length) {
+    return;
+  }
+
+  unsigned int min_x0, min_y0;
+  if (dist > 0.0) {
+    // Adjust starting point and offset to start from min_length distance
+    min_x0 = (unsigned int)(x0 + dx_full / dist * min_length);
+    min_y0 = (unsigned int)(y0 + dy_full / dist * min_length);
+  } else {
+    // dist can be 0 if [x0, y0]==[x1, y1].
+    // In this case only this cell should be processed.
+    min_x0 = x0;
+    min_y0 = y0;
+  }
+  unsigned int offset = min_y0 * size_x + min_x0;
+
+  int dx = x1 - min_x0;
+  int dy = y1 - min_y0;
+
+  unsigned int abs_dx = abs(dx);
+  unsigned int abs_dy = abs(dy);
+
+  int offset_dx = dx > 0 ? 1 : -1;            // sign(dx);
+  int offset_dy = dy > 0 ? size_x : -size_x;  // sign(dy) * size_x;
+
+  float scale = (dist == 0.0) ? 1.0 : min(1.f, max_length / dist);
```

Comment:
for safety, how about using epsilon?

```
std::abs(dist) < epsilon
```

Reply:
Addressed in 5c14fc06f8c34cc1f9b27b7dd61ec746cdcce9ce
(this was essentially a copy from the upstream)